### PR TITLE
feat: boundary condition t3d file converter (unst-8672)

### LINKED ISSUE #756: t3d file
## .t3D legacy file format
.t3D files contain timeseries with multiple data columns to provide boundary values for multiple vertical layers on a boundary.
* Example file:
```
LAYER_TYPE=SIGMA
LAYERS=0.0 0.5 0.55 0.75 1
TIME = 0 seconds since 2006-01-01 00:00:00 +00:00
41 36.45455 36 34 31
TIME = 180 seconds since 2006-01-01 00:00:00 +00:00
41.00002 36.45456 36.00002 34.00002 31.00002
[..]
```
* all keywords are case-insensitive.
* Note: this is the legacy file format. It needs a Pydantic class of its own. It contains the same data, but is **not to be confused with** class `T3D`, which is a subpart of .bc files (the new preferred format).

**Header:**
* valid values for `LAYER_TYPE`:
  * `SIGMA`
  * `Z`
* `LAYERS`: list of whitespace-separated floats, the number of layers is determined after parsing and string splitting.
* *optionally* after the `LAYERS` line can be a line:
  `VECTORMAX = <n>`
  with n a positive integer, defining the number of vector components. Default (when absent) value is 1 (scalar timeseries for each later). Sometimes a velocity vector is specified using `VECTORMAX = 2`.
  This ends the header.

**Data/body:**
Now the data part of the file starts. Each record consists of two lines:
* `TIME = [..] ` where time format is `<float> <time unit> since <reference date time>`
  * `<time unit>` is one of: 'seconds', 'minutes', 'hours', 'days'
  * `<reference date time>` is of ISO 8601 format:
    * `YYYY-MM-DD`,
    * followed by optional time `hh:mm:ss` (prefixed by either 'T' or ' '), default '00:00:00'
    * followed by optional time zone `hh:mm` (prefixed by either 'Z' or '+' or '-')

### PR DESCRIPTION
**Pull request by MAfarrag**
# New Feature T3D Converter
- detailed description of the t3d file is descripted in the issue #756 

# T3D submodule
- A new submodule in the `dflowfm` module is created to handle the t3d IO operations (parser/serializer)
- the `models` module has `T3DTimeRecord` , `T3DModel`. classes to store all the content of the .t3d file

### `T3DTimeRecord` 
- The `T3DTimeRecord`  is created to store the timestamp and the values at this timestamp.
- The `T3DTimeRecord`  validates the format of the reference data given in each record.
```
TIME = 0 seconds since 2006-01-01 00:00:00 +00:00
40 35 34.5 32.5 30
```
### `T3DModel`
- The `T3DModel` class is created to store a list of `T3DTimeRecord`  and the rest of data in the file
- The model does not allow different size `T3DTimeRecord` 
```
LAYER_TYPE=SIGMA
LAYERS=0.0 0.5 0.55 0.75 1
+ List [`T3DTimeRecord` ]
```
###  `T3DParser`
- The `T3DParser` parser have the following assumptions
- There could be comments at the top of the file (similar to the tim file) PLEASE IF THIS IS NOT APPLICABLE TELL ME TO DELETE
- The parser also assumes that it might be comments in the middle of the data, and for that the parser will raise an error  PLEASE IF THIS IS NOT APPLICABLE TELL ME TO DELETE

### `T3DSerializer`
- serializer to write the content of the `T3DModel` class to desk.

# `converters`
## T3DToForcingConverter
- A new converter is created to convert the `T3DModel` to a `T3D` forcing object that can be stored in the .bc file.
- The converter loops over a list of given `T3DModel` objects and returns a list of `T3D` forcing list that can be inserted as a forcing the the `ForcingModel` that represent the `.bc` file.
- The Converter assumes that 

### STILL MISSING
- the way how to detect the type of the vertical position data in the .t3d file is still not integrated, so the following lines are hard coded.
```
vertpositiontype=percBed
```

## `TimToForcingConverter`
- The `TimToForcingConverter` is adjusted to return a list of forcing instead of returning a `ForcingModel`

# Tests
### **TestT3DModelAttributes**
- **`test_layer_type`**: Verifies that `LayerType` correctly initializes from a string.

### **TestT3DTimeRecord**
- **`test_time_record_initialization`**: Tests various initialization scenarios of `T3DTimeRecord`, including valid and invalid time formats.
- **`test_properties`**: Checks if `T3DTimeRecord` properties (`unit`, `time_offset`, `reference_date`) are correctly derived from the time string.

### **TestT3DModel**
- **`test_model_initialization`**: Ensures `T3DModel` initializes correctly with valid quantity names and raises an error for mismatched dimensions.
- **`test_get_units`**: Verifies that `T3DModel.get_units()` returns the expected unit list for different quantity names.
- **`test_different_record_length`**: Ensures a `ValidationError` is raised when records have inconsistent lengths.
- **`test_initialize_with_dict`**: Confirms that `T3DModel` initializes correctly from a dictionary.
- **`test_initialize_with_file_path`**: Checks if `T3DModel` loads correctly from a file path.
- **`test_as_dict`**: Validates that `T3DModel.as_dict()` returns the correct dictionary structure.

### **TestParser**
- **`test_parse_t3d_files`**: Ensures `T3DParser` correctly parses a `.t3d` file and extracts expected data fields.
- **`test_end_of_file_error`**: Checks that the parser handles files with empty lines at the end without errors.

### **TestT3DModelSerializer**
- **`test_serialize_data`**: Ensures `T3DSerializer.serialize()` correctly writes `.t3d` files matching reference files.
- **`test_save_model`**: Verifies that calling `T3DModel.save()` correctly saves the model to a `.t3d` file.

## **test_converter_boundary_condition.py**
- The test module contains all the tests related to the `main_converter` and the boundary condition factory converter.
- The test is refactored to remove the duplication 

## **test_converter_source_sink.py**
- The test module contains all the tests related to the `main_converter` and the boundary condition factory converter.
- The test is refactored to remove the duplication 
 
## **test_t3d_to_bc.py**
- Test the stand alone `T3DModel` to `T3D` converter.

# Example of a final conversion 

```
LAYER_TYPE=SIGMA
LAYERS=0.0 0.5 0.55 0.75 1
TIME = 0 seconds since 2006-01-01 00:00:00 +00:00
40 35 34.5 32.5 30
TIME = 180 seconds since 2006-01-01 00:00:00 +00:00
80 35 34.5 32.5 30
TIME = 9999999 seconds since 2006-01-01 00:00:00 +00:00
40 35 34.5 32.5 30
```
```
[Forcing]
name              = L1_0001
function          = t3d
offset            = 0.0
factor            = 1.0
vertPositions     = 0.0 0.5 0.55 0.75 1.0
vertInterpolation = linear
vertPositionType  = percBed
timeInterpolation = linear
quantity          = time
unit              = seconds since 2006-01-01 00:00:00 +00:00
quantity          = waterlevelbnd
unit              = m
vertPositionIndex = 1
quantity          = waterlevelbnd
unit              = m
vertPositionIndex = 2
quantity          = waterlevelbnd
unit              = m
vertPositionIndex = 3
quantity          = waterlevelbnd
unit              = m
vertPositionIndex = 4
quantity          = waterlevelbnd
unit              = m
vertPositionIndex = 5
0.0        40.0  35.0  34.5  32.5  30.0
180.0      80.0  35.0  34.5  32.5  30.0
9999999.0  40.0  35.0  34.5  32.5  30.0
```

# Issues 
- closes #756

### CODE REVIEW

File: hydrolib/core/dflowfm/t3d/models.py [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+        ['m3/s', 'm', '1e-3', 'degC']
+
+        ```
+    """
+    # Define the mapping of keywords to units
+    unit_mapping = {
+        "discharge": "m3/s",
+        "waterlevel": "m",
+        "salinity": "1e-3",
+        "temperature": "degC",
```
Comment (arthurvd) [inline hydrolib/core/dflowfm/t3d/models.py:305]: This may be correct, but I wonder: where is the change made from "salinitybnd" to "salinity"?
And: does the resulting .bc file still contain `Quantity = salinitybnd`, etc?

File: hydrolib/core/dflowfm/t3d/models.py [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+        """
+        if not TIME_PATTERN.match(value):
+            raise ValueError(
+                f"Time string '{value}' is not in the expected format:\n"
+                "    <float> <time unit> since <ISO date> [optional time] [optional timezone]"
+            )
+        return value
+
+    @property
+    def unit(self) -> str:
```
[the remark this answers is not in the document]
Comment (MAfarrag) [reply hydrolib/core/dflowfm/t3d/models.py:97]: renamed

File: hydrolib/core/dflowfm/t3d/models.py [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+
+    ```
+    """
+
+    comments: List[str] = Field(default_factory=list)
+    records: List[T3DTimeRecord] = Field(default_factory=list)
+    layers: List[float] = Field(default_factory=list)
+    vectormax: Optional[int] = Field(default=1, alias="VECTORMAX")
+    layer_type: LayerType = Field(default=None, alias="LAYER_TYPE")
+    quantities_names: Optional[List[str]] = Field(default=None)
```
[the remark this answers is not in the document]
Comment (MAfarrag) [reply hydrolib/core/dflowfm/t3d/models.py:188]: added to the docstring and an example

File: hydrolib/core/dflowfm/t3d/models.py (hunk 1 above)
Comment (MAfarrag) [reply hydrolib/core/dflowfm/t3d/models.py:318]: well the function checks id the word `salinity` is part of the quantity name, so `anytextsalinity`, `salinityanytext`, `anytextsalinityanytext` will be assigned `1e-3`

File: hydrolib/core/dflowfm/t3d/models.py [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+        (2, 4)
+
+    ```
+    """
+
+    comments: List[str] = Field(default_factory=list)
+    records: List[T3DTimeRecord] = Field(default_factory=list)
+    layers: List[float] = Field(default_factory=list)
+    vectormax: Optional[int] = Field(default=1, alias="VECTORMAX")
+    layer_type: LayerType = Field(default=None, alias="LAYER_TYPE")
```
[the remark this answers is not in the document]
Comment (MAfarrag) [reply hydrolib/core/dflowfm/t3d/models.py:187]: should i create an issue for this?

[the remark this answers is not in the document]
Comment (arthurvd) [reply hydrolib/core/dflowfm/t3d/models.py:187]: Yes, would be great, thanks.

**Review by arthurvd — APPROVED** [review]
Looks good now with yesterday's changes.
